### PR TITLE
Use arch-specific versions of inflate_fast.

### PR DESCRIPTION
**Pull request by nmoinvaz**
This should reduce the cost of indirection that occurs when calling functable
chunk copying functions inside inflate_fast. It should also allow the compiler
to optimize the inflate fast path for the specific architecture.

Background: 
- https://twitter.com/kungfuj35u5/status/1562484460226232321
- https://twitter.com/nmnvzr/status/1563746385799036929

### CODE REVIEW

**Comment by nmoinvaz** [comment]
# Test 1

## Silesia-small 

### HEAD https://github.com/zlib-ng/zlib-ng/commit/89763032d57e3da5301f4cd6e1f363e7a1f85f02
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inf-develop/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.015/0.015/0.016/0.000        0.014/0.015/0.015/0.000       15,737,543
 1     54.166%      0.087/0.088/0.090/0.001        0.044/0.045/0.045/0.000        8,523,732
 2     43.871%      0.145/0.148/0.155/0.002        0.044/0.047/0.052/0.002        6,903,609
 3     42.387%      0.200/0.205/0.215/0.005        0.044/0.045/0.047/0.001        6,670,099
 4     41.647%      0.220/0.226/0.234/0.004        0.043/0.044/0.050/0.002        6,553,723
 5     41.216%      0.230/0.232/0.234/0.001        0.041/0.043/0.044/0.001        6,485,938
 6     41.037%      0.264/0.267/0.270/0.002        0.041/0.043/0.045/0.001        6,457,776
 7     40.778%      0.333/0.337/0.341/0.002        0.042/0.044/0.045/0.001        6,416,919
 8     40.704%      0.423/0.426/0.432/0.003        0.043/0.044/0.045/0.001        6,405,244
 9     40.409%      0.500/0.507/0.513/0.004        0.041/0.043/0.045/0.001        6,358,951

 avg1  48.622%                        0.245                          0.041
 avg2  54.025%                        0.272                          0.046
 tot                                 73.560                         12.372       76,513,534
```
### Branch https://github.com/nmoinvaz/zlib-ng/commit/c2559754f94a178cd1d73256d7f7a46f160cf03a
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.015/0.016/0.018/0.001        0.014/0.016/0.020/0.002       15,737,543
 1     54.166%      0.089/0.090/0.092/0.001        0.042/0.043/0.044/0.001        8,523,732
 2     43.871%      0.147/0.149/0.150/0.001        0.044/0.049/0.061/0.007        6,903,609
 3     42.387%      0.199/0.207/0.217/0.006        0.042/0.044/0.047/0.001        6,670,099
 4     41.647%      0.221/0.226/0.229/0.002        0.041/0.043/0.044/0.001        6,553,723
 5     41.216%      0.231/0.235/0.239/0.002        0.041/0.042/0.043/0.001        6,485,938
 6     41.037%      0.267/0.270/0.276/0.002        0.041/0.043/0.044/0.001        6,457,776
 7     40.778%      0.337/0.341/0.345/0.003        0.040/0.043/0.043/0.001        6,416,919
 8     40.704%      0.426/0.429/0.432/0.002        0.041/0.044/0.054/0.004        6,405,244
 9     40.409%      0.503/0.509/0.516/0.004        0.041/0.042/0.044/0.001        6,358,951

 avg1  48.622%                        0.247                          0.041
 avg2  54.025%                        0.275                          0.045
 tot                                 74.115                         12.277       76,513,534
```
About 0.7% decompression speed improvement.

## Silesia

### HEAD https://github.com/zlib-ng/zlib-ng/commit/89763032d57e3da5301f4cd6e1f363e7a1f85f02
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inf-develop/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.150/0.153/0.155/0.001        0.133/0.138/0.139/0.002      211,973,953
 1     44.409%      0.929/0.938/0.942/0.003        0.474/0.494/0.497/0.005       94,127,290
 2     35.518%      1.559/1.592/1.600/0.010        0.499/0.506/0.512/0.004       75,282,961
 3     33.882%      2.174/2.219/2.227/0.013        0.473/0.484/0.489/0.004       71,816,478
 4     33.174%      2.420/2.472/2.484/0.017        0.468/0.473/0.475/0.002       70,315,668
 5     32.660%      2.528/2.582/2.594/0.019        0.459/0.471/0.474/0.004       69,225,542
 6     32.508%      2.931/3.006/3.017/0.019        0.460/0.466/0.467/0.002       68,902,222
 7     32.255%      3.917/4.013/4.027/0.021        0.457/0.467/0.469/0.003       68,366,800
 8     32.167%      6.189/6.310/6.336/0.029        0.457/0.467/0.469/0.002       68,180,776
 9     31.887%      6.653/6.794/6.813/0.028        0.456/0.462/0.463/0.001       67,586,442

 avg1  40.847%                        3.008                          0.443
 avg2  45.385%                        3.342                          0.492
 tot                                902.353                        132.781      865,778,132
```
### Branch https://github.com/nmoinvaz/zlib-ng/commit/c2559754f94a178cd1d73256d7f7a46f160cf03a
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inffast/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.151/0.155/0.157/0.002        0.137/0.139/0.140/0.001      211,973,953
 1     44.409%      0.930/0.935/0.939/0.003        0.456/0.465/0.470/0.004       94,127,290
 2     35.518%      1.576/1.585/1.592/0.004        0.477/0.483/0.487/0.003       75,282,961
 3     33.882%      2.194/2.205/2.213/0.005        0.455/0.461/0.465/0.003       71,816,478
 4     33.174%      2.435/2.444/2.451/0.005        0.448/0.452/0.455/0.002       70,315,668
 5     32.660%      2.571/2.583/2.593/0.005        0.445/0.450/0.453/0.002       69,225,542
 6     32.508%      2.993/3.004/3.013/0.005        0.442/0.445/0.449/0.002       68,902,222
 7     32.255%      3.939/3.954/3.966/0.007        0.443/0.448/0.451/0.002       68,366,800
 8     32.167%      6.190/6.239/6.260/0.019        0.444/0.446/0.449/0.002       68,180,776
 9     31.887%      6.654/6.688/6.700/0.011        0.436/0.438/0.440/0.001       67,586,442

 avg1  40.847%                        2.979                          0.423
 avg2  45.385%                        3.310                          0.470
 tot                                893.724                        126.787      865,778,132
```
About 4.51% decompression speed improvement.

**Comment by nmoinvaz** [comment]
# Test 2

## Silesia-small

### HEAD https://github.com/zlib-ng/zlib-ng/commit/89763032d57e3da5301f4cd6e1f363e7a1f85f02
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inf-develop/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.016/0.016/0.017/0.000        0.015/0.016/0.017/0.001       15,737,543
 1     54.166%      0.088/0.090/0.093/0.001        0.045/0.046/0.048/0.001        8,523,732
 2     43.871%      0.146/0.155/0.163/0.006        0.046/0.047/0.048/0.001        6,903,609
 3     42.387%      0.201/0.210/0.217/0.005        0.043/0.046/0.049/0.001        6,670,099
 4     41.647%      0.225/0.232/0.240/0.006        0.043/0.044/0.046/0.001        6,553,723
 5     41.216%      0.231/0.238/0.246/0.004        0.043/0.044/0.045/0.001        6,485,938
 6     41.037%      0.268/0.274/0.282/0.004        0.042/0.046/0.058/0.004        6,457,776
 7     40.778%      0.339/0.344/0.350/0.004        0.042/0.045/0.054/0.002        6,416,919
 8     40.704%      0.426/0.437/0.448/0.007        0.043/0.045/0.048/0.001        6,405,244
 9     40.409%      0.508/0.516/0.524/0.005        0.042/0.046/0.057/0.004        6,358,951

 avg1  48.622%                        0.251                          0.043
 avg2  54.025%                        0.279                          0.047
 tot                                 75.409                         12.768       76,513,534
```
### Branch https://github.com/nmoinvaz/zlib-ng/commit/c2559754f94a178cd1d73256d7f7a46f160cf03a
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inffast/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.016/0.017/0.017/0.000        0.015/0.017/0.025/0.003       15,737,543
 1     54.166%      0.088/0.089/0.092/0.001        0.043/0.044/0.046/0.001        8,523,732
 2     43.871%      0.146/0.150/0.157/0.004        0.044/0.048/0.058/0.004        6,903,609
 3     42.387%      0.200/0.207/0.215/0.004        0.042/0.044/0.045/0.001        6,670,099
 4     41.647%      0.220/0.224/0.227/0.002        0.042/0.044/0.049/0.002        6,553,723
 5     41.216%      0.231/0.237/0.243/0.004        0.041/0.043/0.044/0.001        6,485,938
 6     41.037%      0.266/0.271/0.274/0.002        0.040/0.043/0.047/0.002        6,457,776
 7     40.778%      0.331/0.339/0.346/0.004        0.042/0.044/0.046/0.001        6,416,919
 8     40.704%      0.422/0.430/0.437/0.004        0.041/0.045/0.055/0.004        6,405,244
 9     40.409%      0.499/0.509/0.519/0.005        0.041/0.043/0.047/0.001        6,358,951

 avg1  48.622%                        0.247                          0.042
 avg2  54.025%                        0.275                          0.046
 tot                                 74.199                         12.457       76,513,534
```
About 2.43% speed improvement.

## Silesia

### HEAD https://github.com/zlib-ng/zlib-ng/commit/89763032d57e3da5301f4cd6e1f363e7a1f85f02
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inf-develop/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.148/0.156/0.160/0.003        0.132/0.139/0.143/0.003      211,973,953
 1     44.409%      0.927/0.978/1.018/0.037        0.480/0.511/0.527/0.016       94,127,290
 2     35.518%      1.565/1.657/1.734/0.070        0.496/0.525/0.544/0.017       75,282,961
 3     33.882%      2.185/2.317/2.424/0.093        0.472/0.508/0.521/0.013       71,816,478
 4     33.174%      2.441/2.582/2.706/0.104        0.466/0.490/0.505/0.012       70,315,668
 5     32.660%      2.537/2.695/2.825/0.119        0.457/0.488/0.503/0.015       69,225,542
 6     32.508%      2.958/3.129/3.284/0.131        0.454/0.487/0.501/0.014       68,902,222
 7     32.255%      3.956/4.178/4.380/0.174        0.459/0.482/0.499/0.015       68,366,800
 8     32.167%      6.210/6.526/6.865/0.279        0.457/0.481/0.497/0.016       68,180,776
 9     31.887%      6.707/7.060/7.408/0.298        0.454/0.480/0.493/0.013       67,586,442

 avg1  40.847%                        3.128                          0.459
 avg2  45.385%                        3.476                          0.510
 tot                                938.401                        137.688      865,778,132
```
### Branch https://github.com/nmoinvaz/zlib-ng/commit/c2559754f94a178cd1d73256d7f7a46f160cf03a
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inffast/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.145/0.148/0.150/0.001        0.132/0.141/0.153/0.007      211,973,953
 1     44.409%      0.912/0.928/0.935/0.005        0.447/0.461/0.469/0.006       94,127,290
 2     35.518%      1.561/1.586/1.599/0.010        0.472/0.485/0.496/0.007       75,282,961
 3     33.882%      2.200/2.214/2.226/0.008        0.452/0.463/0.478/0.008       71,816,478
 4     33.174%      2.415/2.448/2.461/0.012        0.438/0.453/0.465/0.007       70,315,668
 5     32.660%      2.550/2.594/2.607/0.014        0.442/0.452/0.461/0.006       69,225,542
 6     32.508%      2.959/3.011/3.033/0.020        0.432/0.446/0.455/0.006       68,902,222
 7     32.255%      3.920/3.977/4.001/0.018        0.432/0.449/0.458/0.007       68,366,800
 8     32.167%      6.173/6.204/6.224/0.016        0.434/0.447/0.454/0.006       68,180,776
 9     31.887%      6.594/6.697/6.719/0.027        0.427/0.436/0.447/0.005       67,586,442

 avg1  40.847%                        2.981                          0.423
 avg2  45.385%                        3.312                          0.470
 tot                                894.204                        126.984      865,778,132
```
About 7.77% decompression speed improvement.

**Comment by nmoinvaz** [comment]
As far as code size goes, I am not seeing any notable differences between HEAD and this PR.

**Comment by nmoinvaz** [comment]
Very strange big endian failures..

**Comment by mtl1979** [comment]
The MinGW i686 failure seems unrelated, but the other failures need to be investigated...

**Comment by nmoinvaz** [comment]
My benchmarks are based on the first two commits. I am not seeing any real performance benefit to making the chunk copy functions `static inline` in commit 3.

**Comment by mtl1979** [comment]
> My benchmarks are based on the first two commits. I am not seeing any real performance benefit to making the chunk copy functions `static inline` in commit 3.

Using `static inline` just avoids duplicate symbols when linking, it doesn't change the actual code.

**Comment by KungFuJesus** [comment]
So I did measure quite a bit of difference, now that things are inlined, in removing this specific fast path from the AVX version:
https://github.com/zlib-ng/zlib-ng/blob/cd8b6eb8de425c1608153cb6399d4839a547257f/chunkset_tpl.h#L111

Now that we're not having to eat the cost of a call, this no longer translates to an early jump to another function with the same prologue and epilogue.  

It's debatable whether or not we want to add back a fast path there that incorporates the length and only does a single shuffle.  Given that it might translate to just two instructions after the jump - it may still be worthwhile (there's an obvious early exit for the first 128 bit vector, just cast it with undefined bits).

I also, under the speculation from @dougallj, unconditionally unrolled the copy loop afterward by 3 in chunk_memset, overlapped or not, hoping the uarch's store forwarding was reasonable.  It seemed to be a minor performance win (it's unclear if the store forwarding accounted for the win or if the collapsed branch and reduced instructions did).  His testing seemed to indicate it had been beneficial for aarch64 as well.

**Comment by KungFuJesus** [comment]
> As far as code size goes, I am not seeing any notable differences between HEAD and this PR.

_Somewhat_ surprising, we are wholesale duplicating the inffast decode loop.  Which, while a tighter loop with fewer branches, is still a fair bit of code.  The bodies of code we are inlining into the inffast functions are very short - you might see some shrinkage there if the function setup and tear down is girthier than using the register resident variables as-is.

**Comment by nmoinvaz** [comment]
The 3rd commit only adds about 1-2KB code size on x86.

**Comment by nmoinvaz** [comment]
The code is a lot more complex because we are using `sse2` chunk functions in `sse41`. It means I have to have two copies of those functions (one for the other implementation which is `Z_INTERNAL`) and (one for `inflate_fast` which is `static inline`).

Even if we got rid of that though, I'd still have to have `CSTATIC` macro for `CHUNKMEMSET_SAFE` which is also exported in `functable`.

But you are right, we are essentially duplicating the `inflate_fast` loop but also duplicating all the `chunk` functions quite a bit.

File: chunkset_tpl.h [hunk 1]
```diff
@@ -5,6 +5,9 @@
 #include "zbuild.h"
 #include <stdlib.h>
 
+#define MAKE_CSTATIC(name, suffix) name ## suffix
+#define CSTATIC(name)              MAKE_CSTATIC(name, _static)
+
 #if CHUNK_SIZE == 32 && defined(X86_SSE41) && defined(X86_SSE2)
```
Comment (KungFuJesus) [inline chunkset_tpl.h]: For now it's faster to actually remove this.  Maybe we can hack in support for a sub-chunk size (halfchunk_t?) and skip the blend that has to occur between the two halves is dist is < 32.  

Edit: Ugh, meant lines 11 down.  Github's ability to select lines is pretty limited.

File: chunkset_tpl.h [hunk 2]
```diff
@@ -168,7 +177,11 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
     return out;
```
Comment (KungFuJesus) [inline chunkset_tpl.h]: Something in these inlined functions appears to be emitting several memcpy calls, I'm wonder if it's the lines above.  It is guaranteed to be a sub vector size so it's going to be only a small handful of move instructions.  It also could be happening in the chunkmemset_# functions, though I would expect the compiler would try to optimize those away.

**Review by KungFuJesus — CHANGES_REQUESTED** [review]
Other than my couple of comments this looks fairly clean.  It claws back as much as 7% of the delta between us and @dougallj 's cloudflare fork (still about another 10% not accounted for).  @dougallj 's optimizations that aren't yet integrated could bring that a bit closer.

One significant trade off is we're conceding memcpy in places that are extremely beneficial for huge run lengths where we can copy out a solid couple of KB, and ERMS aware memcpy implementations shine.  For shorter copy sequences this could be hurting us.  Deriving a simple heuristic to determine whether or not to actually call memcpy versus expand a copy of loop might be feasible but could be treacherous.  Cloudflare seems to be invoking the compiler builtin rather than the library call basically everywhere in their chunkcopy routines.

File: chunkset_tpl.h (hunk 1 above)
Comment (nmoinvaz) [reply chunkset_tpl.h]: I have removed that code in my latest commit.

**Comment by KungFuJesus** [comment]
So I'm not seeing significant differences with and without the inlining happening on the chunk functions.  However, the compiler _is_ inlining chunkmemset for me now with that commit where it wasn't before (but it for certain looks like chunkcopy was being inlined).  I'd have to do wider testing but I would have expected repeatedly calling into chunkmemset would have been more expensive.  It might be the opposite, at least on my benchmark PNG image I'm testing with.  It's probably going to be very data dependent, I imagine.  

So far I think it's still compelling to head this direction, I might see if I can find the cause of those call memcpy sequences and play with compiler builtins or copy loops in place of them if they are short.

**Comment by mtl1979** [comment]
`static inline` means the function can't be called from other compilation units, so compiler is more inclined to inline the function without creating non-inlined copy. Even without `inline` keyword, compiler is free to inline the function, but always have to emit non-inlined copy.

**Comment by nmoinvaz** [comment]
> So I'm not seeing significant differences with and without the inlining happening on the chunk functions.

This confirms what I was seeing as well. Here were my benchmarks just for posterity.

### HEAD https://github.com/zlib-ng/zlib-ng/commit/89763032d57e3da5301f4cd6e1f363e7a1f85f02
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inf-develop/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.165/0.167/0.169/0.001        0.148/0.152/0.154/0.001      211,973,953
 1     44.409%      1.042/1.052/1.056/0.003        0.543/0.551/0.556/0.003       94,127,290
 2     35.518%      1.784/1.791/1.799/0.004        0.563/0.570/0.575/0.003       75,282,961
 3     33.882%      2.496/2.508/2.515/0.005        0.537/0.543/0.547/0.003       71,816,478
 4     33.174%      2.792/2.803/2.809/0.004        0.527/0.533/0.538/0.003       70,315,668
 5     32.660%      2.909/2.925/2.935/0.006        0.523/0.530/0.537/0.004       69,225,542
 6     32.508%      3.385/3.399/3.406/0.006        0.516/0.524/0.527/0.002       68,902,222
 7     32.255%      4.528/4.544/4.554/0.007        0.518/0.525/0.529/0.003       68,366,800
 8     32.167%      6.997/7.028/7.040/0.010        0.516/0.523/0.526/0.002       68,180,776
 9     31.887%      7.616/7.630/7.639/0.006        0.513/0.520/0.525/0.004       67,586,442

 avg1  40.847%                        3.385                          0.497
 avg2  45.385%                        3.761                          0.552
 tot                               1015.436                        149.109      865,778,132
```
### PR https://github.com/zlib-ng/zlib-ng/pull/1337/commits/cd8b6eb8de425c1608153cb6399d4839a547257f
```
OS: Windows 10 10.0.22000 AMD64
CPU: Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
Tool: ../zlib-ng/build-inffast/Release/minigzip.exe
Levels: 0-9
Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 0    100.008%      0.164/0.167/0.169/0.001        0.150/0.153/0.155/0.001      211,973,953
 1     44.409%      1.037/1.043/1.047/0.003        0.523/0.534/0.541/0.005       94,127,290
 2     35.518%      1.779/1.790/1.795/0.004        0.547/0.555/0.560/0.004       75,282,961
 3     33.882%      2.494/2.502/2.509/0.005        0.524/0.531/0.535/0.003       71,816,478
 4     33.174%      2.783/2.789/2.794/0.004        0.510/0.520/0.524/0.003       70,315,668
 5     32.660%      2.934/2.944/2.952/0.005        0.509/0.516/0.520/0.003       69,225,542
 6     32.508%      3.410/3.421/3.430/0.005        0.508/0.513/0.519/0.003       68,902,222
 7     32.255%      4.517/4.532/4.538/0.005        0.506/0.513/0.518/0.003       68,366,800
 8     32.167%      6.995/7.016/7.028/0.008        0.506/0.512/0.516/0.002       68,180,776
 9     31.887%      7.571/7.594/7.603/0.009        0.501/0.508/0.514/0.003       67,586,442

 avg1  40.847%                        3.380                          0.486
 avg2  45.385%                        3.755                          0.539
 tot                               1013.937                        145.659      865,778,132
```

About 2.3% decompression improvement which is not as great as the first two commits. In all my tests I am using MSVC so it might be that MSVC is having a harder time with inlining. The inlining might not make such a difference because now all the functions `chunk*` and `inflate_fast` as in the same compilation unit and might get placed in the binary closer together?

I think it would be better if this PR was more limited in scope. It will allow us to better benchmark the chunkcopy inlining in a different PR and check the performance difference. For that reason I am going to limit this PR only to using arch-specific versions of `inflate_fast`. 

I have backed up the old changes I had here in my `functable-inffast-v2` branch https://github.com/nmoinvaz/zlib-ng/tree/improvements/functable-inffast-v2.

**Comment by nmoinvaz** [comment]
> So far I think it's still compelling to head this direction, I might see if I can find the cause of those call memcpy sequences and play with compiler builtins or copy loops in place of them if they are short.

That sounds like a good avenue of investigation. I think we should at some point add chunkcopy functions to the Google benchmarks.

**Comment by KungFuJesus** [comment]
> `static inline` means the function can't be called from other compilation units, so compiler is more inclined to inline the function without creating non-inlined copy. Even without `inline` keyword, compiler is free to inline the function, but always have to emit non-inlined copy.

Right, it's a good sledgehammer way to ensure the compiler truly does inline your function properly.  Our intuition that the optimizer found it attractive enough to inline the chunk copy routines jives with benchmarks.  It's possible we could compact the chunkmemset well enough that it becomes beneficial to inline that, as well.

**Review by KungFuJesus — APPROVED** [review]
Looks good to me

**Comment by mtl1979** [comment]
> Right, it's a good sledgehammer way to ensure the compiler truly does inline your function properly. Our intuition that the optimizer found it attractive enough to inline the chunk copy routines jives with benchmarks. It's possible we could compact the chunkmemset well enough that it becomes beneficial to inline that, as well.

Any code can be simplified, but there is always the risk that it becomes slower than the longer, more complex, version. We also need to keep the code readable for other developers/contributors, otherwise it will be hard to work on the code without breaking it by editing the surrounding lines.

Just by looking at benchmark results is not enough to warrant changing code as benchmarks can have wide error margin due to the test environment being far from normal running environment. We can't benchmark on every possible CPU and memory configuration.

**Comment by KungFuJesus** [comment]
> Any code can be simplified, but there is always the risk that it becomes slower than the longer, more complex, version. We also need to keep the code readable for other developers/contributors, otherwise it will be hard to work on the code without breaking it by editing the surrounding lines.

I agree with that sentiment but I don't believe the changes proposed in this PR make the code less maintainable or more complex.  If anything, it might make things a little less complicated, as we have far fewer dispatch points to the function table. Apart from some name mangling through macros to get to the specific function, I don't think that things are made any more complicated by this.  This PR is also pretty careful to only dispatch where things differ in the copy functions rather than going hogwild and creating one inffast per specialization that we have on _any_ function.  

There's also an implicit benefit here to the people compiling with -march=generic (such as a distro maintainer) in that they finally have access to BMI2 instructions with a dynamic dispatch where it matters most (this loop).  That across the board buys some gains that @nmoinvaz might be seeing in this.

**Comment by mtl1979** [comment]
@KungFuJesus I was just implying that we shouldn't change this PR more than was needed to make the CI pass... Obviously some failures were unrelated to this PR and need to be resolved in separate PRs.

**Comment by nmoinvaz** [comment]
@Dead2 have you had opportunity to look at this PR?

**Comment by Dead2** [comment]
> @Dead2 have you had opportunity to look at this PR?

I have looked at it, and I like what I see, but I have not yet done any testing.

I finally finished the most pressing outdoor projects 2 days before Christmas, so hopefully that means I have more spare the next few months. Just have to get through the Christmas and New Years party season first 😄.

**Comment by nmoinvaz** [comment]
Rebased.

**Comment by nmoinvaz** [comment]
Bump.

**Review by Dead2 — APPROVED** [review]

**Comment by Dead2** [comment]
This needs a little rebase now

**Comment by nmoinvaz** [comment]
Rebased.

**Pull request merged**